manual/LowLevel/Functions.md

```diff
@@ -193,13 +217,225 @@ apply1_union :: Apply1Union
+myCallback = ProgressUpdate_Deref $ \progress ->
+  putStrLn $ "Progress: " ++ show progress ++ "%"
+
+-- Preferred: automatic cleanup with withToFunPtr
+withToFunPtr myCallback $ \funPtr -> do
+  onProgressChanged (ProgressUpdate funPtr)
+
+-- Or manually manage the function pointer lifetime with bracket
+bracket
+  (toFunPtr myCallback)
+  (freeHaskellFunPtr . un_ProgressUpdate)
+  (\funPtr -> onProgressChanged (ProgressUpdate funPtr))
+```
+
+#### Unwrapping function pointers
+
+To call a function pointer returned from C, use `fromFunPtr`:
+
+```hs
+do
+  validatorFunPtr <- getValidator
+  -- validatorFunPtr :: DataValidator
+
+  -- Extract the FunPtr and convert to Haskell function
+  let validator = fromFunPtr (un_DataValidator validatorFunPtr)
+  result <- un_DataValidator_Deref validator 42
+```
+
+### Example: struct with function pointer fields
+
+Function pointers frequently appear as struct fields for registering handlers:
+
+```c
+struct MeasurementHandler {
+  void (*onReceived)(struct Measurement *data);
+  int (*validate)(struct Measurement *data);
+  void (*onError)(int errorCode);
+};
+
+void registerHandler(struct MeasurementHandler *handler);
+```
+
+In Haskell we can make use of the `ToFunPtr` to construct the
+`MeasurementHandler` record.
+
+```hs
+alloca $ \handlerPtr -> do
+  onReceivedPtr <- toFunPtr $ OnReceived_Deref $ \dataPtr -> do
+    measurement <- peek dataPtr
+    print measurement
+
+  validatePtr <- toFunPtr $ Validate_Deref $ \dataPtr -> do
+    -- validation logic
+    return 1
+
+  onErrorPtr <- toFunPtr $ OnError_Deref $ \errorCode ->
+    putStrLn $ "Error: " ++ show errorCode
+
+  poke handlerPtr $ MeasurementHandler
+    { measurementHandler_onReceived = onReceivedPtr
+    , measurementHandler_validate = validatePtr
+    , measurementHandler_onError = onErrorPtr
+    }
+
+  registerHandler handlerPtr
+```
+
 ## Userland CAPI
 
-TODO
+GHC's foreign function interface has limitations on which C functions can be
+imported directly. For example, GHC cannot import functions that take or return
+structs by value. To work around these limitations, `hs-bindgen` generates C
+wrapper functions that can be imported by GHC, and then generates Haskell
+bindings to these wrappers instead of to the original C functions.
+
+This approach is similar to GHC's `capi` calling convention, which also
+generates C wrappers to handle features that the FFI cannot express directly.
+However, by generating these wrappers ourselves at the userland level, we can
+extend the set of supported function signatures beyond what GHC's `capi`
+provides. For instance, we can handle by-value struct arguments and return
+values, which `capi` does not support.
+
+The generated wrappers are named with a prefix based on the Haskell module
+name. For example, a C function `foo` might have a wrapper `B_foo` that we
+import instead:
+
+```c
+// Generated C wrapper
+void B_print_point ( struct point * arg1 ) {
+  print_point ( arg1 );
+}
```

Comment:
That's not right, is it..? We use some kind of hash, to avoid name collisions?

Reply:
right, I was taking this from the paper, but that's no longer true